### PR TITLE
Removes forcePut in exchange for Optional UserPermission, which is em…

### PR DESCRIPTION
**Pull request by ttomsu**
…pty in case any resource provider is unavailable.

@jtk54 @duftler PTAL.

tag: #29

### CODE REVIEW

File: fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java [hunk 1]
```diff
@@ -48,15 +46,7 @@
...
-  private AccountProvider accountProvider;
-
-  @Autowired
-  @Setter
-  private ApplicationProvider applicationProvider;
-
-  @Autowired
-  @Setter
-  private ServiceAccountProvider serviceAccountProvider;
+  private List<ResourceProvider> resourceProviders;
```
Comment (jtk54) [inline fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java]: Nice

File: fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java [hunk 2]
```diff
@@ -67,8 +57,8 @@
     if (!anonymousEnabled) {
       return Optional.empty();
     }
-    return Optional.of(getUserPermission(AnonymousUserConfig.ANONYMOUS_USERNAME,
-                                         new ArrayList<>(0) /* groups */));
+    return getUserPermission(AnonymousUserConfig.ANONYMOUS_USERNAME,
+                             new ArrayList<>(0) /* groups */);
```
Comment (jtk54) [inline fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java]: Why is the `<>` not filled in? The groups will be `String`s, right?

File: fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java [hunk 3]
```diff
@@ -103,32 +93,26 @@
...
+    for (ResourceProvider provider : resourceProviders) {
       try {
-        permission.setServiceAccounts(serviceAccountProvider.getAll(groups));
+        permission.addResources(provider.getAll(groups));
       } catch (ProviderException pe) {
-        permission.setPartialPermission(true);
+        log.warn("Can't resolve permission for '%s' due to error with '%s'",
+                 userId,
+                 provider.getClass().getSimpleName());
+        log.debug("Error resolving UserPermission", pe);
```
Comment (jtk54) [inline fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java]: I think at least the short message/reason would be worth logging in `warn`.

**Comment by jtk54** [comment]
LGTM after comments.

File: fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java (hunk 2 above)
Comment (ttomsu) [inline fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java]: It's a feature of Java 7 - the diamond operator is used by the compiler to infer the type.
 
http://www.javaworld.com/article/2074080/core-java/jdk-7--the-diamond-operator.html

File: fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java (hunk 3 above)
Comment (ttomsu) [inline fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/DefaultPermissionsResolver.java]: Good call. Done.

**Pull request merged**